HTTP client: simplify and improve HTTP(S) proxy support

DDvO opened the pull request:
This fixes minor bugs and handles a design flaw with the `bio_update_fn` connect/disconnect callback function of type `OSSL_HTTP_bio_cb_t` not having a parameter of type `OSSL_HTTP_REQ_CTX *`, which causes trouble with connecting via an HTTPS proxy during TLS connection setup because this needs access to the proxy and no_proxy information. Moreover, the callback needs access to the `SSL_CTX` and to the server hostname for setting the Server Name Indication (SNI). All of this has been done so far in a pretty cumbersome and not entirely correct way.

~Make sure that `OSSL_HTTP_proxy_connect()` is called if and only if needed with TLS.
   So far, this did not correctly take into account the environment variables `https_proxy` and `no_proxy`.
   Doing this fix unfortunately requires generalizing the callback function such that it can take into account whether a proxy is actually being used during connection setup.
   This is done in a binary backward compatible way: by adding to `OSSL_HTTP_bio_cb_t` an `rtcx` parameter that can be used unless the `details` parameter is `1` on connect, which is tried first for compatibility with OpenSSL 3.0.
 The generalization also makes the design of `OSSL_HTTP_bio_cb_t` more future-proof.~

~Also introduce `OSSL_HTTP_REQ_CTX_proxy_connect()`, an improved variant of `OSSL_HTTP_proxy_connect()`, which strongly simplifies the use of the callback function, e.g., in `apps/`.~

**Update of Feb 9, 2026:** 
I recently found a way of working around the  missing `OSSL_HTTP_REQ_CTX` pointer that does not require adding a parameter to `OSSL_HTTP_bio_cb_t`, which would have incurred an API break.
This is possible by introducing `OSSL_HTTP_REQ_CTX_proxy_connect()`, a variant of `OSSL_HTTP_proxy_connect()`
 that uses a `OSSL_HTTP_REQ_CTX` pointer, and calling this function (in case an SSL/TLS connection is being opened)
already from `OSSL_HTTP_open()`, where the pointer is available, rather than letting the callback do this.
Moreover, using `SSL_CTX_{set,get}_ex_data()` to convey the server host name via the `SSL_CTX`
to the callback function avoids having to pass the server name via the generic callback arg.
As a result, the callback arg can be reduced to the bare minimum: a pointer to the `SSL_CTX`.

Implementing this improvement I noticed that that userinfo (user name and password) that can be provided with the proxy URI was not used, neither for the HTTP nor the HTTPS proxy case. I added this. 

This PR also includes several fixes:
* added several missing failure checks to `OSSL_HTTP_proxy_connect()`
* fix `OSSL_HTTP_open()` to include in the `Host:` header line the server port, which is needed for non-default ports
* fix the CMP app to provide the right proxy usage info

* ~Fix cleanup of TLS BIO via `bio_update_fn` callback function.
   Make `app_http_tls_cb()` tidy up on disconnect the SSL BIO it pushes on connect.
   Make `OSSL_HTTP_close()` respect this.~

* ~`OSSL_HTTP_proxy_connect()`: Fix glitch in response HTTP header parsing.~

~It would be good to have CI tests for using an HTTP(S) proxy, but this would require providing within OpenSSL a test proxy that supports HTTP and HTTPS connections.~
This PR meanwhile contains client-side unit tests for the provided improvements of HTTP(S) proxy use,
which has become possible after lifting some not really needed argument restrictions of `OSSL_HTTP_open()`.
I tested the improved implementation also with external both HTTP and HTTPS proxies, including proxy client credential (user name and password) usage.

~This PR also includes a commit that fixes an omission documenting the `ok` parameter of `OSSL_HTTP_close()`.~

DDvO commented:
Finally this PR is ready for review.
It was pretty hard to come up with a good solution for those bugs that is backward compatible with the high-level HTTP client API recently introduced for 3.0.

t8m commented:
I am sorry, but I do not really see this being merged to 3.0 branch. But perhaps some other OTC members might have different opinion.

DDvO commented:
Thanks for your initial comment. Why don't you see this possible for the 3.0 branch?
I propose inclusion in the 3.0 branch for two reasons.
* These are bug fixes, which unfortunately are not as small as one would wish; otherwise the HTTP client cannot be properly used with HTTPS.
* This PR also fixes a design flaw that I unfortunately did not realize before 3.0 was released. The earlier it is fixed, the less chance is that applications use it (in comparison with the improved version).

t8m commented:
> Thanks for your initial comment. Why don't you see this possible for the 3.0 branch? I propose inclusion in the 3.0 branch for two reasons.
> 
>     * These are bug fixes, which unfortunately are not as small as one would wish due a subtle issue of the design of the rather young high-level HTTP client API; otherwise the HTTP client cannot be properly used with HTTPS.
> 
>     * This PR also fixes a design flaw that I unfortunately did not realize before 3.0 was released. The earlier it is fixed, the less chance is that applications use it (in comparison with the improved version).

This is completely contrary to the stable release policy proposal. Once some API is released, we have to live with that until the next major version.

DDvO commented:
Hmm, I spent quite some effort to make this change binary backward compatible.
And shouldn't it be possible to add functions?

t8m commented:
> This PR also fixes a design flaw that I unfortunately did not realize before 3.0 was released. The earlier it is fixed, the less chance is that applications use it (in comparison with the improved version).

This sounds like an API break to me.

DDvO commented:
> > This PR also fixes a design flaw that I unfortunately did not realize before 3.0 was released. The earlier it is fixed, the less chance is that applications use it (in comparison with the improved version).
> 
> This sounds like an API break to me.

My point with this was that, as as always, 
the earlier a bug fix becomes usable, the less chance of misbehavior.
With the type `OSSL_HTTP_bio_cb_t`:
```
BIO *(*)(BIO *bio, void *arg, int connect, int detail);
```
getting extended to
```
BIO *(*)(BIO *bio, void *arg, int connect, int detail, OSSL_HTTP_REQ_CTX *rctx);
```
applications get the chance to use the corrected HTTPS proxy behavior. This is the bug fix aspect.

Moreover, using this, the new function `OSSL_HTTP_REQ_CTX_proxy_connect()` makes the use of the connect/disconnect callback function much simpler, such that setting up, filling, and using an auxiliary structure, as currently done in `apps/`:
```
typedef struct app_http_tls_info_st {
    const char *server;
    const char *port;
    int use_proxy;
    long timeout;
    SSL_CTX *ssl_ctx;
} APP_HTTP_TLS_INFO;
```
is no more needed.

So in both these senses (bug fix and ease of use) it is very beneficial to take advantage of the improvement as early as possible.

DDvO commented:
An API break would mean that something can go wrong due to a version mismatch.
Yet the way I designed this extension avoids this.
Or can you point out a concrete use case where due to some version mismatch a crash or wrong behavior can occur?

DDvO commented:
In my view, although this PR adapts the public header `http.h`,
it qualifies as a bug fix, according to the following definition from https://github.com/openssl/technical-policies/pull/8/files#diff-04f7314e4a706d785d4bc8420b6d5390c07dc15365affede06f372583faf8d4f:

> A **bug fix** is a fix of functionality of the libraries, modules,
> applications, or the build system that (all or any items might apply):
> 
> - makes the functionality conform with the existing end-user documentation
>  where such a change does not break existing common usage
> - makes the end-user documentation conform with the existing behavior of the
>  software

In case you do not agree, please let me know exactly why,
such that we can sort this out, 
either by clarification or by any needed reduction of this PR.

t8m commented:
I might not understand fully the intent and implementation of this fix. So I am going to abstain from reviewing it. I am not going to block its inclusion in 3.0.

t8m commented:
A change of API signature breaks existing code that is going to be compiled. That is by definition an API break.

t-j-h commented:
It is an API change - when you change the number of args in a callback you are changing the API contract. That cannot be done outside of a major release. Inside a minor this requires a new function (a new callback).

DDvO committed:
HTTP client: simplify workaround for OSSL_HTTP_bio_cb_t not having a OSSL_HTTP_REQ_CTX* parameter

Reduce the arg type of the bio_update_fn used by APPS to its bare minimum: a SSL_CTX pointer.
This is possible by
1. introducing OSSL_HTTP_REQ_CTX_proxy_connect(), a variant of OSSL_HTTP_proxy_connect()
that uses a OSSL_HTTP_REQ_CTX pointer, and calling this in case a SSL/TLS connection is being opened
already from OSSL_HTTP_open() (where the pointer is available) rather than letting the callback do this.
2. using SSL_CTX_{set,get}_ex_data to convey the server host info via the SSL_CTX to the callback function.

DDvO commented:
Yes, it is an API change, thus formally an API break, Yet as I tried to point out,
* I carefully designed the solution in a way that re-compilation of existing applications is not needed and
* tested that indeed the previous version of the `apps/` runs fine when dynamically linked with the new crypto lib including the fix.

So I still believe that semantically it is not an API break.

Anyway, to avoid spending more time on discussing this,
I give up on fixing this issue before the next major release.
I'm going to come up with a workaround or point this out as a known bug.

t8m commented:
It is an API break if you need to change your code to be able to compile again. Period. It might not be an ABI break if what you say about your solution is true.

mattcaswell commented:
OTC: This is 4.0 only

DDvO commented:
I suppose the OTC hold can be lifted,
given the decision that this targets 4.0.

DDvO commented:
Since this PR was decided to target 4.0,
is there anything that can/should be done on it in the near future apart from occasionally rebasing it?

DDvO committed:
HTTP client: take userinfo in proxy URL as HTTP(S) proxy client credentials

DDvO committed:
fix OSSL_HTTP_open() to check non-NULL result of http_req_ctx_new() and to include the Host: header line first, adding non-default server port

DDvO committed:
APPS/cmp: fix info on HTTP(S) proxy use: HTTP vs. HTTPS proxy and do not show any given user:pass@ part

DDvO committed:
HTTP client: show server and proxy diagonstic info also on early connection error

t8m commented:
Good work. With the recent changes this can be approved I think. Would you please add some tests?

DDvO committed:
OSSL_HTTP_transfer.pod: fix doc of OSSL_HTTP_open() regarding use of |bio| and |rbio|

DDvO committed:
relax OSSL_HTTP_open() parameter requirements and use this to strengthen HTTP(S) proxy use test cases

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

DDvO commented:
rebased to fix new merge conflicts

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 30 days ago

openssl-machine commented:
This PR is in a state where it requires action by @openssl/committers but the last update was 61 days ago

andrewkdinh commented:
Hi, please rebase this PR again. Then I will do a review of it

DDvO commented:
> Hi, please rebase this PR again. Then I will do a review of it

Thank you @andrewkdinh for your reminder to rebase this - done now.
Very glad that you are going to review this PR!

andrewkdinh commented on apps/lib/apps.c:
```diff
@@ -2980,36 +2980,54 @@ static BIO *http_tls_shutdown(BIO *bio)
     return bio;
 }
 
+static int app_tls_server_ex_data_idx = -1;
```
Probably not a real issue because this should pretty much always be single-threaded, but we could consider using `CRYPTO_THREAD_run_once` here just in case

DDvO replied:
Yes, all our apps (in particular the CMP app, which makes direct use of `app_SSL_CTX_set_tls_server_ex_data()`), so no real issue.

andrewkdinh reviewed (dismissed):
LGTM

DDvO commented:
Rebased to fix (pretty trivial) merge conflicts

DDvO commented:
Thanks @andrewkdinh for your approval of end-July.
As I meanwhile had to rebase due to merge conflicts, I need to ask for re-approval.

idrassi commented on crypto/http/http_client.c:
```diff
@@ -1100,17 +1162,17 @@ OSSL_HTTP_REQ_CTX *OSSL_HTTP_open(const char *server, const char *port,
         return NULL;
     }
-    if (rbio != NULL && (bio == NULL || bio_update_fn != NULL)) {
+    if (bio == NULL && rbio != NULL) {
```
I don't think this relaxation is safe for TLS. `bio_update_fn` only receives `cbio`, so it has no way through this API to replace a separate `rctx->rbio`. HTTPS response parsing continues to read directly from the supplied `rbio`.

There's a second problem in the same case: `OSSL_HTTP_REQ_CTX_proxy_connect` uses `rctx->wbio` for both sides of CONNECT. The new test puts the fake reply into `wbio`, so it doesn't exercise the documented role of `rbio` as the response BIO. With a real split transport, the reply arrives on `rbio` and CONNECT can't complete.

I would keep the old restriction unless the callback and CONNECT paths both handle the read BIO.

idrassi commented on crypto/http/http_client.c:
```diff
@@ -1100,17 +1162,17 @@ OSSL_HTTP_REQ_CTX *OSSL_HTTP_open(const char *server, const char *port,
 
+    if (port != NULL && *port == '\0')
+        port = NULL;
+    proxy = OSSL_HTTP_adapt_proxy(proxy, no_proxy, server, use_ssl);
```
This looks blocking to me. `OSSL_HTTP_adapt_proxy` is making a routing choice before we know whether OpenSSL owns the connection. When `bio` is supplied, its peer has already been chosen, so an environment proxy can't redirect it.

I reproduced this with `http_proxy=http://corpuser:corppass@proxy.invalid:8080` and a caller supplied memory BIO. The request written to that BIO included:

```text
GET http://origin.example/resource HTTP/1.0
Proxy-Authorization: Basic Y29ycHVzZXI6Y29ycHBhc3M=
```

That sends credentials from the process environment to a peer selected by the caller.

I think we should keep environment proxy adaptation inside the `bio == NULL` path. If an explicit proxy string is meant to describe a supplied BIO, that path shouldn't consult the environment or `no_proxy`.

idrassi commented on crypto/http/http_client.c:
```diff
@@ -1137,31 +1196,43 @@ OSSL_HTTP_REQ_CTX *OSSL_HTTP_open(const char *server, const char *port,
         if (bio == NULL) /* cbio was not provided by caller */
-            BIO_free_all(cbio);
-        goto end;
+            BIO_free(cbio);
```
There's an allocation failure double free here. `http_req_ctx_new` sets `free_wbio` before it duplicates any strings. If a later duplication fails, its error path frees `rctx`, which has already freed `cbio`. This branch then calls `BIO_free(cbio)` again.

Setting `rctx->free_wbio = free_wbio` only after the string duplications succeed would keep `cbio` owned by the caller on every constructor failure.

idrassi commented on crypto/http/http_client.c:
```diff
@@ -1608,3 +1690,48 @@ int OSSL_HTTP_proxy_connect(BIO *bio, const char *server, const char *port,
+            pass = colon + 1;
+        }
+    }
+    ret = OSSL_HTTP_proxy_connect(rctx->wbio, rctx->server, rctx->port,
```
`port == NULL` means default to 443 in `OSSL_HTTP_proxy_connect`. The issue is that `OSSL_HTTP_open` also accepts an explicit service embedded in `server` when the separate `port` argument is NULL. For a direct connection, `http_new_bio` preserves the combined host and service for `BIO_new_connect`.

When a proxy is selected, the target value is copied unchanged into `rctx`. Passing it directly here loses the distinction between no service and a service embedded in `server`.

For `server = "origin.example:8443"` and `port = NULL`, the emitted request is:

```text
CONNECT [origin.example:8443]:443 HTTP/1.0
```

The colon is mistaken for a naked IPv6 address, then the default port is appended.

Could `OSSL_HTTP_open` normalize the server and port before storing them, or could this wrapper split `rctx->server` before calling `OSSL_HTTP_proxy_connect`? If the combined form is not intended with proxies, `OSSL_HTTP_open` should reject it explicitly rather than emit this CONNECT request.

idrassi commented on include/openssl/http.h:
```diff
@@ -79,6 +79,9 @@ OSSL_HTTP_REQ_CTX *OSSL_HTTP_open(const char *server, const char *port,
 int OSSL_HTTP_proxy_connect(BIO *bio, const char *server, const char *port,
     const char *proxyuser, const char *proxypass,
     int timeout, BIO *bio_err, const char *prog);
+int OSSL_HTTP_REQ_CTX_proxy_connect(OSSL_HTTP_REQ_CTX *rctx,
+    const char *proxyuser, const char *proxypass,
+    BIO *bio_err, const char *prog);
```
I don't see the intended public call sequence for this function.
`OSSL_HTTP_open` calls it before returning `rctx`. On a 407 response, `OSSL_HTTP_open` frees the context, so the caller can't retry with `proxyuser` and `proxypass`. After success, calling this public function would send a second CONNECT inside the established tunnel.

The public constructor can't populate the private proxy fields either. I think this should stay internal unless the opening API lets an application provide credentials before CONNECT.

idrassi commented on apps/cmp.c:
```diff
@@ -2412,21 +2426,14 @@ static int setup_client_ctx(OSSL_CMP_CTX *ctx)
-        info->timeout = OSSL_CMP_CTX_get_option(ctx, OSSL_CMP_OPT_MSG_TIMEOUT);
-
-        if (info->ssl_ctx == NULL)
+        if ((ssl_ctx = setup_ssl_ctx(ctx, host)) == NULL)
```
This is a functional regression: `opt_tls_host` is not used here anymore, so the option no longer affects certificate name verification.

I used a certificate for `server.example`. The server was reached at `https://127.0.0.1:19443` with `-tls_host server.example`. Verification still expected `127.0.0.1` and failed with an IP address mismatch.

`setup_ssl_ctx` now needs a verification name distinct from its SNI name. We should pass the override to `truststore_set_host_etc` while keeping the connection host for SNI.

idrassi reviewed (changes requested):
Thanks @DDvO for the extensive work on these improvements.

I have put inline comments on the issues I found while reviewing the changes. I confirmed the runtime findings with HTTP probes and a live CMP TLS check.

I have few other observations, but they can wait until these points are addressed.
